pkg/member/member_add.go

```diff
@@ -0,0 +1,264 @@
+	retryPeriod = 5 * time.Second
+
+	// etcdTimeout is timeout for etcd operations
+	etcdTimeout = 10 * time.Second
+)
+
+// AddMember interface defines the functionalities needed to add a member to the etcd cluster
+type AddMember interface {
+	// AddMemberAsLearner add a new member as a learner to the etcd cluster
+	AddMemberAsLearner(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+	IsMemberInCluster(*logrus.Logger, *brtypes.EtcdConnectionConfig)
+
+	// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed only if its logs are caught up with the leader
+	PromoteMember(context.Context, *logrus.Entry, *brtypes.EtcdConnectionConfig)
+}
+
+// NewMember holds the configuration for the mechanism of adding a new member to the cluster.
+type NewMember struct {
+	clientFactory client.Factory
+}
+
+// NewMemberConfig returns new ExponentialBackoff.
+func NewMemberConfig(etcdConnConfig *brtypes.EtcdConnectionConfig) *NewMember {
+	etcdConn := *etcdConnConfig
+	svcEndpoint, err := miscellaneous.GetEtcdSvcEndpoint()
+	if svcEndpoint == "" || err != nil {
+		svcEndpoint = "http://127.0.0.1:2379"
+	}
+	etcdConn.Endpoints = []string{svcEndpoint}
+	clientFactory := etcdutil.NewFactory(etcdConn)
+	return &NewMember{
+		clientFactory: clientFactory,
+	}
+}
+
+// AddMemberAsLearner add a member as a learner to the etcd cluster
+func (m *NewMember) AddMemberAsLearner(logger *logrus.Logger) error {
+	//Add member as learner to cluster
+	logger.Info("Starting Add Member")
+	memberURL, err := getMemberURL(logger)
+	if memberURL == "" || err != nil {
+		logger.Info("Error fetching etcd member URL")
+		return fmt.Errorf("Could not fetch member URL : %v", err)
+	}
+
+	cli, err := m.clientFactory.NewCluster()
+	defer cli.Close()
+	if err != nil {
+		logger.Info("Error creating etcd client to add member")
+		return fmt.Errorf("failed to build etcd cluster client : %v", err)
+	}
+
+	memAddCtx, cancel := getContext(context.TODO(), etcdTimeout)
+	defer cancel()
+	_, err = cli.MemberAddAsLearner(memAddCtx, []string{memberURL})
+
+	if err == nil {
+		logger.Info("Added member to cluster as a learner")
+		return nil
+	}
+	if errors.Is(err, rpctypes.Error(rpctypes.ErrGRPCPeerURLExist)) {
+		logger.Info("Member already part of etcd cluster")
+		return nil
+	}
+	if err != nil {
+		logger.Warn("Error adding member as a learner: ", err)
+		return fmt.Errorf("Error adding member as a learner: %v", err)
+	}
+
+	logger.Info("Could not as member as learner due to: ", err)
+	return fmt.Errorf("Could not as member as learner due to: %v", err)
+}
+
+// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+func (m *NewMember) IsMemberInCluster(logger *logrus.Logger) (bool, error) {
+	// Check if an etcd is already available
+	backoff := retry.DefaultBackoff
+	backoff.Steps = 2
```

Comment:
are we trying only 2 times here ?

Reply:
This retry is for `ProbeEtcd()`. I think 2 attempts should suffice to see if a cluster is already up or not
wdyt? do you recommend another value here?

Reply:
make sense